File: leetcode/src/02207_maximize_number_of_subsequences_in_a_string/solution.go
package leetcode

func maximumSubsequenceCount(text string, pattern string) int64 {
	t := []rune(text)
	p := []rune(pattern)
	cnt2 := int64(0)
	for _, ch := range t {
		if ch == p[1] {
			cnt2++
		}
	}
	sum := int64(0)
	curSuffix := cnt2
	for _, ch := range t {
		if ch == p[1] {
			curSuffix--
		}
		if ch == p[0] {
			sum += curSuffix
		}
	}
	cnt1 := int64(0)
	extra := cnt2
	for _, ch := range t {
		if ch == p[0] {
			cnt1++
		}
		if ch == p[1] {
			cnt2--
		}
		extra = max(extra, cnt1, cnt2)
	}
	return sum + extra
}

func max(nums ...int64) int64 {
	res := nums[0]
	for _, n := range nums {
		if res < n {
			res = n
		}
	}
	return res
}

File: leetcode/src/02207_maximize_number_of_subsequences_in_a_string/solution.rs
struct Solution;

impl Solution {
    pub fn maximum_subsequence_count(text: String, pattern: String) -> i64 {
        let t: Vec<_> = text.chars().collect();
        let p: Vec<_> = pattern.chars().collect();
        let mut res = 0;
        let (mut cnt1, mut cnt2) = (0, 0);
        for &ch in &t {
            if ch == p[1] {
                res += cnt1;
                cnt2 += 1;
            }
            if ch == p[0] {
                cnt1 += 1;
            }
        }
        res + cnt1.max(cnt2)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let text = String::from("abdcdbc");
        let pattern = String::from("ac");
        assert_eq!(Solution::maximum_subsequence_count(text, pattern), 4);

        let text = String::from("aabb");
        let pattern = String::from("ab");
        assert_eq!(Solution::maximum_subsequence_count(text, pattern), 6);

        let text = String::from("aabaa");
        let pattern = String::from("aa");
        assert_eq!(Solution::maximum_subsequence_count(text, pattern), 10);
    }
}
